sklearn/cluster/k_medoids_.py

```diff
@@ -0,0 +1,284 @@
+                # and assign the newly found medoid as the medoid
+                # for cluster k
+                medoid_idxs[k] = cluster_k_idxs[min_cost_idx]
+
+    def transform(self, X):
+        """Transforms X to cluster-distance space.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = (n_samples, n_features)
+            Data to transform.
+
+        Returns
+        -------
+        X_new : {array-like, sparse matrix}, shape=(n_samples, n_clusters)
+            X transformed in the new space of distances to cluster centers.
+        """
+        X = check_array(X, accept_sparse=['csr', 'csc'])
+        check_is_fitted(self, "cluster_centers_")
+
+        return pairwise_distances(X, Y=self.cluster_centers_,
+                                  metric=self.distance_metric)
+
+    def predict(self, X):
+        """Predict the closest cluster for each sample in X
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = (n_samples, n_features)
+            New data to predict.
+
+        Returns
+        -------
+        labels : array, shape = (n_samples,)
+            Index of the cluster each sample belongs to.
+        """
+        check_is_fitted(self, "cluster_centers_")
+        X = check_array(X, accept_sparse=['csr', 'csc'])
+
+        distances = pairwise_distances(X, Y=self.cluster_centers_,
+                                       metric=self.distance_metric)
+
+        # Assign data points to clusters based on which cluster assignment
+        # yields the smallest distance
+        labels = np.argmin(distances, axis=1)
+
+        return labels
+
+    def _compute_inertia(self, X):
+        """Compute inertia of new samples. Inertia is defined as the sum of the
+        sample distances to closest cluster centers.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = (n_samples, n_features)
+            Samples to compute inertia for.
+
+        Returns
+        -------
+        Sum of sample distances to closest cluster centers.
+        """
+        distances = self.transform(X)
+
+        # Define inertia as the sum of the sample-distances
+        # to closest cluster centers
+        inertia = np.sum(np.min(distances, axis=1))
+
+        return inertia
+
+    def _get_initial_medoid_indices(self, distances, n_clusters):
+        """Select initial mediods randomly or heuristically"""
+
+        if self.init == 'random':  # Random initialization
+            # Pick random k medoids as the initial ones.
+            medoids = self.random_state_.permutation(
+                distances.shape[0])[:n_clusters]
+        elif self.init == 'heuristic':  # Initialization by heuristic
+            # Pick K first data points that have the smallest sum distance
+            # to every other point. These are the initial medoids.
+            medoids = list(np.argsort(np.sum(distances, axis=1))[:n_clusters])
```

Comment:
use `argpartition` to reduce work here. Also do not convert to list; we do not in the random case.

Reply:
Great idea. I didn't think about argpartition. Never used it before, I'm not sure if I got this right:

```
np.argpartition(np.sum(D, axis=1), range(0, n_clusters))[:n_clusters]
```

Is that what you've been thinking of?